java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -35,41 +37,72 @@ import kotlinx.coroutines.withContext
  */
 class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     val storageKey: StorageKey,
-    val backingType: Type,
-    val callbackFactory: (String) -> ProxyCallback<Data, Op, T>
+    val backingType: Type
 ) {
     private val storeMutex = Mutex()
     private val log = TaggedLog { "DirectStoreMuxer" }
+    private val proxyManager = RandomProxyCallbackManager<Data, Op, T>(
+        "backing",
+        Random
+    )
 
     // TODO(b/158262634): Make this CacheMap Weak.
     /* internal */ val stores = LruCacheMap<String, StoreRecord<Data, Op, T>>(
         50,
         livenessPredicate = { _, sr -> !sr.store.closed }
-    ) { _, sr -> closeStore(sr) }
+    ) { muxId, sr -> closeStore(muxId, sr) }
+
+    // Keeps track of all the stores a callbackId is registered to. Used when removing an
+    // observer (off method), as that requires unregistering to each store it is registered with.
+    private val callbackIdToMuxIdMap = mutableMapOf<Int, MutableSet<String>>()
+
+    fun on(callback: ProxyCallback<Data, Op, T>, callbackToken: Int? = null): Int {
+        synchronized(proxyManager) {
+            val callbackId = proxyManager.register(callback, callbackToken)
+            callbackIdToMuxIdMap[callbackId] = mutableSetOf()
+            return callbackId
+        }
+    }
+
+    fun off(callbackToken: Int) {
+        for (muxId in callbackIdToMuxIdMap[callbackToken]!!) {
+            val (idSet, store) = checkNotNull(stores[muxId]) { "store not found" }
+            store.off(callbackToken)
+            idSet.remove(callbackToken)
+        }
+        callbackIdToMuxIdMap.remove(callbackToken)
+        synchronized(proxyManager) {
+            proxyManager.unregister(callbackToken)
+        }
+    }
 
     /** Safely closes a [DirectStore] and cleans up its resources. */
-    private fun closeStore(storeRecord: StoreRecord<*, *, *>) {
+    private fun closeStore(muxId: String, storeRecord: StoreRecord<*, *, *>) {
         if (!storeRecord.store.closed) {
-            log.debug { "close the store(${storeRecord.id})" }
+            log.debug { "close the store($muxId)" }
 
             try {
-                storeRecord.store.close()
+                for (id in storeRecord.idSet) storeRecord.store.off(id)
```

Comment:
Good catch! I added `store.close()` and I also included the removal in `callbackIdToMuxIdMap`.

Calling `store.off(id)` seems to be redundant because `store.close()` unregisters all the callbacks anyways. I still have to loop through to make sure the muxId is removed from each id in `callbackIdToMuxIdMap`. Do you think I should leave it as it (clearer to see we are explicitly unregistering from store) or remove redundancy?